REGRESSION (Safari 26): `@font-face` / `FontFace.family` fails when font family name contains space (e.g. "Lato 2")

darinadler opened the pull request:
#### 3208096991b941e064c217b0a903a8374f2507a2
<pre>
REGRESSION (Safari 26): `@font-face` / `FontFace.family` fails when font family name contains space (e.g. &quot;Lato 2&quot;)
<a href="https://bugs.webkit.org/show_bug.cgi?id=300620">https://bugs.webkit.org/show_bug.cgi?id=300620</a>
<a href="https://rdar.apple.com/162637501">rdar://162637501</a>

Reviewed by Tim Nguyen.

While the specification isn&apos;t clear enough on this, the FontFace interface should
treat the font family as just a string, without parsing or serializing steps.
This is the de facto standard since both Chrome and Firefox work this way, and
WebKit worked roughly this way before changes earlier this year.

Besides this bug fix it would be good to:
- make the specification correct and unambiguous on this point
- add tests to Web Platform Tests
- add tests covering FontFace family setter
- add tests covering FontFace with family from @font-face rule

* LayoutTests/fast/text/font-face-family-expected.txt: Updated to expect font
family name to be treated as a string, not parsed with the &lt;font-family&gt; syntax.
* LayoutTests/fast/text/font-face-family.html: Ditto.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::family const): Get the string value of the font family
property rather than serializing it. Code is slightly confusing because of the
way we share the StyleProperties between the font-family descriptor and the
font-family property, since the latter can contain a list of families, and
can include generic families rather than families specified by font name.

* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::create): Updated since setFamily no longer requires a
ScriptExecutionContext argument since it no longer does any parsing.
(WebCore::FontFace::setFamily): Set from a string, without parsing.
Retains the &quot;empty string is a syntax error&quot; behavior, since we have a
test that expects it; would be easy to remove that too and allow empty
strings, but I&apos;d want to test the other browser engine behavior first,
and it&apos;s not urgent to deal with this edge case.
* Source/WebCore/css/FontFace.h: Removed the ScriptExecutionContext from
the setFamily function, since it no longer does parsing.
* Source/WebCore/css/FontFace.idl: Ditto.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
(WebCore::CSSPropertyParserHelpers::parseFontFaceFontFamily): Deleted.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.h: Deleted
unneeded parseFontFaceFontFamily function declaration.

Canonical link: <a href="https://commits.webkit.org/301793@main">https://commits.webkit.org/301793@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c11c915004dd15fbc9739a06fb812326a902661

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127080 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46716 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37804 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134082 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78642 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e9e98d8b-0fee-49bb-a8dd-e60f8929988d) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128951 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47332 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55242 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/96704 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64743 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bc7c3ddc-6f7d-4cb3-8887-515c45035522) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130028 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37888 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113771 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77214 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/44642cf3-0de8-41b3-ab97-a16203eabd3d) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/36758 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31905 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77474 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/107754 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32255 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136608 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53735 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/41389 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/105221 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54239 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110129 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/104911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50438 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28851 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51275 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19872 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53667 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59540 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52905 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56238 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54663 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/text/font-face-family-expected.txt:
```diff
@@ -1,13 +1,13 @@
+PASS (new FontFace('4a', 'url(garbage.otf)')).family is "4a"
+PASS (new FontFace('4"a', 'url(garbage.otf)')).family is "4\"a"
+FAIL (new FontFace('4\'a', 'url(garbage.otf)')).family should be 4\'a. Was 4'a.
+FAIL (new FontFace('4\'a"b', 'url(garbage.otf)')).family should be 4\'a"b. Was 4'a"b.
```
FAIL seems unexpected here

darinadler replied:
Oops, yes, put wrong expected values in the test!

nt1m commented on LayoutTests/fast/text/font-face-family.html:
```diff
@@ -5,16 +5,16 @@
 </head>
 <body>
 <script>
```
Maybe the test should explain the results as a comment on top of the script, so future folks have a bit more context before modifying the test as [293891@main](https://commits.webkit.org/293891@main) did.

nt1m replied:
Could also switch to `js-test.js` while there :)

nt1m reviewed (approved):
WPTs would be good if you have time, but this PR is fine for the purposes of fixing the escape, thanks!

darinadler committed:
REGRESSION (Safari 26): `@font-face` / `FontFace.family` fails when font family name contains space (e.g. "Lato 2")
https://bugs.webkit.org/show_bug.cgi?id=300620
rdar://162637501

Reviewed by Tim Nguyen.

While the specification isn't clear enough on this, the FontFace interface should
treat the font family as just a string, without parsing or serializing steps.
This is the de facto standard since both Chrome and Firefox work this way, and
WebKit worked roughly this way before changes earlier this year.

Besides this bug fix it would be good to:
- make the specification correct and unambiguous on this point
- add tests to Web Platform Tests
- add tests covering FontFace family setter
- add tests covering FontFace with family from @font-face rule

* LayoutTests/fast/text/font-face-family-expected.txt: Updated to expect font
family name to be treated as a string, not parsed with the <font-family> syntax.
* LayoutTests/fast/text/font-face-family.html: Ditto.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::CSSFontFace::family const): Get the string value of the font family
property rather than serializing it. Code is slightly confusing because of the
way we share the StyleProperties between the font-family descriptor and the
font-family property, since the latter can contain a list of families, and
can include generic families rather than families specified by font name.

* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::create): Updated since setFamily no longer requires a
ScriptExecutionContext argument since it no longer does any parsing.
(WebCore::FontFace::setFamily): Set from a string, without parsing.
Retains the "empty string is a syntax error" behavior, since we have a
test that expects it; would be easy to remove that too and allow empty
strings, but I'd want to test the other browser engine behavior first,
and it's not urgent to deal with this edge case.
* Source/WebCore/css/FontFace.h: Removed the ScriptExecutionContext from
the setFamily function, since it no longer does parsing.
* Source/WebCore/css/FontFace.idl: Ditto.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
(WebCore::CSSPropertyParserHelpers::parseFontFaceFontFamily): Deleted.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.h: Deleted
unneeded parseFontFaceFontFamily function declaration.

Canonical link: https://commits.webkit.org/301793@main
